db/db_impl.cc

```diff
@@ -790,6 +861,22 @@ Status DBImpl::SetDBOptions(
             thread_dump_stats_.reset();
           }
       }
+      if (new_options.stats_persist_period_sec !=
+          mutable_db_options_.stats_persist_period_sec) {
+          if (thread_persist_stats_) {
+            mutex_.Unlock();
+            thread_persist_stats_->cancel();
```

Comment:
Since we have a lot of code with this pattern, how about we extend `RepeatableThread::cancel` with

```
void RepeatableThread::cancel(Mutex* mutex) {
    if (mutex != nullptr) {
        mutex->unlock();
    }
    // actual cancel code
    if (mutex != nullptr) {
        mutex->lock();
    }
}